twitter_api/main.py

```diff
@@ -16,6 +21,158 @@ def before_request():
+        if username and password:
+            user = query_db('select * from user where username=?', [username], one=True)
+
+            # check if the user exists
+            if user:
+                if user['password'] != md5(password).hexdigest():
+                    # wrong password
+                    abort(401)
+
+                # check if they are logged in. don't generate duplicate token
+                token = query_db('select * from auth where user_id=?', [user['id']], one=True)
+
+                if not token:
+                    # generate the token
+                    query_db('insert into auth (user_id, access_token) values(?, ?)',
+                             [user['id'], create_token()])
+                    g.db.commit()
+                    # retrieve token
+                    token = query_db('select * from auth where user_id=?', [user['id']], one=True)
+
+                resp = {"access_token": token['access_token']}
+                resp = Response(response=json.dumps(resp), status=201, mimetype="application/json")
+                return resp
+            else:
+                # user not found
+                abort(404)
+    # user did not provide username and password
+    abort(400)
+
+
+@app.route('/logout', methods=['POST'])
+@auth_only
+def logout(*args, **kwargs):
+    access_token = request.get_json()['access_token'] if request.get_json() else abort(401)
+    query_db('delete from auth where access_token=?', [access_token])
+    g.db.commit()
+    return Response(response='', status=204)
+
+
+@app.route('/profile/<username>')
+def get_profile(username):
+    profile = query_db('select id as user_id, username, first_name, last_name, birth_date '
+                       'from user where username=?',
+                       [username],
+                       one=True)
+    if profile:
+        tweets = query_db('select id, content as text, created as date from tweet where user_id=?',
+                          [profile['user_id']])
+        for t in tweets:
+            t['uri'] = '/tweet/{}'.format(t['id'])
+            created_date = datetime.strptime(t['date'], '%Y-%m-%d %H:%M:%S')
+            t['date'] = created_date.strftime("%Y-%m-%dT%H:%M:%S")
+        profile['tweet_count'] = len(tweets)
+        profile['tweets'] = tweets
+
+        return Response(response=json.dumps(profile), status=200, mimetype="application/json")
+
+    return abort(404)
+
+
+@app.route('/profile', methods=['POST'])
+@auth_only
+@json_only
+def post_profile():
+    required_attrs = ['first_name', 'last_name', 'birth_date']
+    data = request.get_json()
+    user_id = get_user_id_from_token(data['access_token'])
+    if not user_id:
+        abort(401)
+    if all(attr in data for attr in required_attrs):
+        query_db('update user set first_name=?, last_name=?, birth_date=? where id=?',
+                 (data['first_name'], data['last_name'],
+                  data['birth_date'],
+                  int(user_id)))
```

Comment:
isn't `user_id` an integer by default? or did it get converted to a string somewhere and I missed that?